File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < s.length() ; i++){
            char c = s.charAt(i);
            if(checkOpenBracket(c)) stack.push(c);
            else{
                if(stack.isEmpty() || !checkSimilarBrackets(stack.pop(), c)) return false;
            }
            
        }
        return stack.isEmpty();
    }
    
    private boolean checkOpenBracket(char c){
        return c == '(' || c == '[' || c == '{';
    }
    
    private boolean checkSimilarBrackets(char c, char d){
        if(d == ')') return c == '(';
        else if (d == ']') return c == '[';
        return c == '{';
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()
        for c in s:
            if self.isClosing(c): 
                if len(stack) > 0 and self.closes(stack[-1], c):
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0
                
        
    
    @staticmethod
    def isClosing(s:str) -> bool:
        return s ==']' or s ==')' or s =='}'
    
    @staticmethod
    def closes(s:str, t:str) -> bool:
        return s == '[' and t == ']' or s == '(' and t == ')' or s == '{' and t == '}'
        